linker/js/src/main/scala/org/scalajs/linker/irio/NodeVirtualIRFiles.scala

```diff
@@ -19,77 +19,217 @@ import scala.concurrent._
+      implicit ec: ExecutionContext): Future[NodeScalaJSIRContainer] = {
+    stat(path).map(s => new NodeVirtualScalaJSIRFile(path, relativePath, version(s)))
+  }
+
+  private def fromDirectory(dir: String)(
+      implicit ec: ExecutionContext): Future[Seq[NodeScalaJSIRContainer]] = {
+    cbFuture[js.Array[FS.Dirent]](FS.readdir(dir, ReadDirOpt, _)).flatMap { entries =>
+      val (dirs, files) = entries.toSeq.partition(_.isDirectory)
+
+      val subdirFiles = Future.traverse(dirs) { e =>
+        val path = Path.join(dir, e.name)
+        fromDirectory(path)
+      }
+
+      // Since we will remove relativePath (#3580) we do not bother calculating it here.
+      val irFileNames = files.map(_.name).filter(_.endsWith(".sjsir"))
+      val directFiles = Future.traverse(irFileNames) { name =>
+        val path = Path.join(dir, name)
+        fromSingleFile(path, "dummy")
+      }
+
+      for {
+        sdf <- subdirFiles
+        df <- directFiles
+      } yield sdf.flatten ++ df
     }
+  }
 
-    VirtualScalaJSIRFile.withPathExceptionContext(path, Future(blocking(read())))
+  private def stat(path: String)(implicit ec: ExecutionContext): Future[FS.Stats] =
+    cbFuture[FS.Stats](FS.stat(path, _))
+
+  private def version(stats: FS.Stats): Option[String] =
+    stats.mtime.map(_.getTime.toString).toOption
+
+  private def isNotFound(e: js.Error): Boolean = {
+    val code = e.asInstanceOf[js.Dynamic].code.asInstanceOf[js.UndefOr[String]]
+    code.exists(_ == "ENOENT")
   }
 }
 
-private[scalajs] class NodeVirtualJarScalaJSIRContainer(val path: String) extends ScalaJSIRContainer {
-  import NodeVirtualJarScalaJSIRContainer.JSZip
+private class NodeVirtualScalaJSIRFile(
+    val path: String,
+    val relativePath: String,
+    val version: Option[String]
+) extends VirtualScalaJSIRFile with NodeScalaJSIRContainer {
+  import NodeInterop._
+
+  def entryPointsInfo(implicit ec: ExecutionContext): Future[ir.EntryPointsInfo] = {
+    val result = cbFuture[Int](FS.open(path, "r", _)).flatMap { fd =>
+      def read(buf: ByteBuffer): Future[Unit] = {
+        val len = buf.remaining()
+        val off = buf.position()
 
-  val version: Option[String] =
-    NodeFS.statSync(path).mtime.map(_.getTime.toString).toOption
+        cbFuture[Int](FS.read(fd, buf.typedArray(), off, len, off, _))
+          .map(bytesRead => buf.position(buf.position() + bytesRead))
+      }
 
-  def sjsirFiles(implicit ec: ExecutionContext): Future[List[VirtualScalaJSIRFile]] = Future {
-    val zip = new JSZip(NodeFS.readFileSync(path))
+      ir.Serializers.deserializeEntryPointsInfo(ByteBuffer.allocateDirect, read)
+    }
+
+    VirtualScalaJSIRFile.withPathExceptionContext(path, result)
+  }
 
+  def tree(implicit ec: ExecutionContext): Future[ir.Trees.ClassDef] = {
+    val result = cbFuture[js.Array[Int]](FS.readFile(path, _)).map { arr =>
+      val stream = new ArrayBufferInputStream(new Uint8Array(arr).buffer)
```

Comment:
This could use a `ByteBuffer` instead, constructed with `TypedArrayBuffer.wrap`.